DeviceListener: Handle non-gesture warmth increment in the native scale

NiLuJe opened the pull request:
Re: #11220
Followup to #6468

NiLuJe commented:
Ping @zwim for more details about the maths in the warmth scale, because I clearly did not understand how it was generated (what's `n`?).

NiLuJe commented:
> Ping @zwim for more details about the maths in the warmth scale, because I clearly did not understand how it was generated (what's `n`?).

By which I mean, I get *vaguely* similar numbers with something like

```
> for i = 0, 10 do n = (math.sqrt(2^i)); print(n) end
1
1.4142135623731
2
2.8284271247462
4
5.6568542494924
8
11.313708498985
16
22.62741699797
32
```

But that doesn't appear to match exactly; and doesn't answer how the cutoff and/or the amount of values being left is chosen?

NiLuJe commented:
Thanks, I'll give it a try later tonight ;).

NiLuJe commented:
Works like a charm, thanks @zwim ;).

NiLuJe commented:
(Rebase me)

NiLuJe reviewed (commented):
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11226)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

hius07 commented:
Maybe change `min` to 0? For frontlight and warmth, now we cannot return to "gesture distance" once the fixed step is set (actually we can for the frontlight via the default action, but not directly via the SpinWidget).
https://github.com/koreader/koreader/blob/0d2f8fbb3fc62a8542f483831867e9ba19950d5c/frontend/dispatcher.lua#L101
Is it info text good?
https://github.com/koreader/koreader/blob/0d2f8fbb3fc62a8542f483831867e9ba19950d5c/frontend/dispatcher.lua#L828

NiLuJe commented:
> Maybe change `min` to 0?

Yup, makes sense.

(untested, gtg)

NiLuJe commented:
> > Maybe change `min` to 0?
> 
> Yup, makes sense.

FWIW, a better approach would be to enforce min > 0 for gestures types that do not have a distance (e.g. tap, hold), because right now that opens up a footgun: setting this to 0 on a tap will do nothing ^^.

NiLuJe commented:
Also, my crappy `min` check is probably stupid, because, IIRC, we have ways of breaking the min/max constraints in a SpinWidget.

hius07 commented:
> IIRC, we have ways of breaking the min/max constraints in a SpinWidget.

How this? I cannot see it in the numberpicker widget.

NiLuJe commented:
> How this? I cannot see it in the numberpicker widget.

By prefixing the input w/ `:` (c.f., https://github.com/koreader/koreader/pull/9182)

NiLuJe commented:
> EDIT: Hmm, in the specific case of fl/warmth, the backend might be doing something decently smart in this case, actually. Will test tonight.

Yup, it falls back to 1, so, that makes it behave as if it were set to 1, we're good ;p.

hius07 commented:
> By prefixing the input w/ `:`

I see, overlooked it. In debug mode only.

NiLuJe commented:
Tweaked the last two commits:

* Implemented @zwim's latest simplification
* Removed the funky conditional that I'd previously added to the SpinWidget help text

NiLuJe reviewed (commented):
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11226)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, @poire-z, and @zwim)



<!-- Sent from Reviewable.io -->

poire-z reviewed (approved):
Approving the things I can. (Can't say anything about the warm maths.)

pazos commented on frontend/device/cervantes/powerd.lua:
```diff
@@ -51,10 +51,13 @@ function CervantesPowerD:init()
                     self.device.frontlight_settings[key] = val
                 end
             end
+            -- Does this device's NaturalLight use a custom scale?
```
Not sure anymore :p

Here you are: https://github.com/koreader/koreader/issues/4291#issuecomment-439700537

I hope it helps :)

NiLuJe replied:
Yup, I'm familiar with that PWM controller, it's on quite a few NTX boards ;).

Frenzie commented on frontend/dispatcher.lua:
```diff
@@ -98,11 +98,11 @@ local settingsList = {
     set_frontlight_warmth = {category="absolutenumber", event="SetFlWarmth", min=0, max=100, title=_("Set frontlight warmth"), screen=true, condition=Device:hasNaturalLight()},
-    increase_frontlight_warmth = {category="incrementalnumber", event="IncreaseFlWarmth", min=1, max=Device:getPowerDevice().fl_warmth_max, title=_("Increase frontlight warmth"), screen=true, condition=Device:hasNaturalLight()},
-    decrease_frontlight_warmth = {category="incrementalnumber", event="DecreaseFlWarmth", min=1, max=Device:getPowerDevice().fl_warmth_max, title=_("Decrease frontlight warmth"), screen=true, condition=Device:hasNaturalLight(), separator=true},
+    increase_frontlight_warmth = {category="incrementalnumber", event="IncreaseFlWarmth", min=0, max=Device:getPowerDevice().fl_warmth_max, title=_("Increase frontlight warmth"), screen=true, condition=Device:hasNaturalLight()},
```
This is all safe? (Wrt to the min level vs off stuff.)

NiLuJe replied:
Yup, there's no notion of "off" for the warmth on any relevant devices (barring the Kobo Aura One where you can't decouple warmth from intensity), and the frontlight backend already allows turning it off, so there's no behavior change.

I had a similar reaction at first, but the value is only relevant as an increment itself, it's not an actual clamp on the effective frontlight warmth (unlike for `absolutenumber` actions).

Frenzie commented on frontend/ui/widget/sortwidget.lua:
```diff
@@ -408,7 +408,8 @@ function SortWidget:_populateItems()
     end
-    self.footer_page:setText(T(_("Page %1 of %2"), self.show_page, self.pages), self.footer_center_width)
+    -- NOTE: We forgo our usual "Page x of y" wording because of space constraints given the way the widget is currently built
+    self.footer_page:setText(T(_("%1 / %2"), self.show_page, self.pages), self.footer_center_width)
```
That probably requires a context since it could also mean completely different things.

NiLuJe replied:
Done, under `Pagination`, as I didn't find anything obviously relevant, and it could come in handy down the line, so I went fairly generic.

NiLuJe reviewed (commented):
> Reviewed 4 of 4 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11226)* status: all files reviewed, 4 unresolved discussions (waiting on @Frenzie, @pazos, and @zwim)



<!-- Sent from Reviewable.io -->

mergen3107 commented:
Hi @NiLuJe !
I tried this PR's diff in its current state and I must say... it is absolutely **stunning**!
Thank you so much for your thoughtful approach.

Warmth:
1) Warmth increments upon gestures (swipes) are now in "native" scale, from 0 to 24. This is now consistent with intensity scale.
2) Increment of `1` is truly increment of 1! Which is great.
3) Increment of `0` (new one) is expectedly scaled with the length of the gesture swipe. I am able to increase/decrease either intensity or warmth by exactly half (12/24) by doing a full-height swipe along the edge (regardless of landscape/portait orientation).

Reading progress: The gap is gone, yay!

Sort widget:
- In `Sort` (I tested in sort gestures' quick menu items) it works as expected.
- Do you guys want to extent this to History/Favorites and File manager lists as well? It still says `Page X of Y` there. I don't have personal opinion on these
- I feel like `Page 1/1` is a little wordy (in case where there is only one page). Don't you guys feel that `Page 1.` instead, or simply dropping the whole row with buttons and page indication is better?

Frenzie commented:
> or simply dropping the whole row with buttons and page indication is better?

I think it's generally clearer to have things there and disabled. But the word "page" is taken out here iirc?

mergen3107 commented:
> But the word "page" is taken out here iirc?

Ah, yes, sorry, missed that! It is just `X/Y` now (which is very clean, btw!)

pazos commented:
Fix #8957

NiLuJe commented:
> Fix #8957

Does it? I haven't touched the startup code, for the reason we discussed: the Kobo implementation is fairly gnarly and Nickel-specific; so, while this shouldn't be terribly hard since it'd be much simpler, someone with a device will have to deal with it because I don't feel comfortable doing that sort of stuff without being able to test it as I go ;).

pazos commented:
Ah, then nope :p 

I misunderstood you.

Most likely in https://www.mobileread.com/forums/showpost.php?p=4380859&postcount=49 when I saw a cervantes code path on your commit.
